packages/hub/writers.js

```diff
@@ -78,21 +82,21 @@ class Writers {
     }
   }
 
-  async _finalizeAndReply(pending) {
+  async _finalizeAndReply(pending, branch, session, type, schema) {
     let meta = await pending.finalize();
+    let contentType = schema.types.get(type);
     let finalDocument = pending.finalDocument;
-    let responseDocument = {
+    finalDocument.meta = meta;
+
+    return new DocumentContext({
+      type,
+      branch,
+      schema,
       id: finalDocument.id,
-      type: finalDocument.type,
-      meta
-    };
-    if (finalDocument.attributes) {
-      responseDocument.attributes = finalDocument.attributes;
-    }
-    if (finalDocument.relationships) {
-      responseDocument.relationships = finalDocument.relationships;
-    }
-    return responseDocument;
+      upstreamDoc: finalDocument,
+      sourceId: contentType.dataSource.id,
+      read: (type, id) => this.searchers.get(session, branch, type, id)
```

Comment:
@ef4 this is how i'm looking up relationships, but this relies on the fact that the writers update the index as evidenced by the `pgseach/indexer: indexes a belongs-to` test. As this create https://github.com/cardstack/cardstack/blob/master/packages/pgsearch/node-tests/indexer-test.js#L107-L112 would need to be indexed before this create https://github.com/cardstack/cardstack/blob/master/packages/pgsearch/node-tests/indexer-test.js#L114-L122 in order for the searcher to find the related record

Reply:
Using a searcher here is correct, although we're going to need to think about permissions. When you save the searchDoc into the search index, it should be complete, so it shouldn't depend on the user's current permissions when following relationships. But when we reply to the client with the pristine doc, it should obviously only include things they're allowed to see.

Reply:
cool, make sense. I'll use the `Session.INTERNAL_PRIVILEGED`

Reply:
or wait, i just parsed your last sentence.... yes, i'll figure out a badd way to refactor this.

Reply:
i think when I get to the jsonapi middleware, i'll revisit this again, but with pristineDocs as input instead of the upstreamDoc, and i think that is probably where we need to use the user's session for the reads

Reply:
or maybe there is a `userPristineDoc()` function on document-context which uses user's session? I'm seeing that the jsonapi middleware just returns the response from teh writers directly.

Reply:
Something like [applyReadAuthorization](https://github.com/cardstack/cardstack/blob/cc237da819a33c9c50eb5f8d059c6adafaa4f8f3/packages/hub/searchers.js#L57) may be the right thing to use for the writer reply.